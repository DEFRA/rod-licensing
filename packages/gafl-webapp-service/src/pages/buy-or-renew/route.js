import { BUY_OR_RENEW } from '../../uri.js'
import Joi from 'joi'
import pageRoute from '../../routes/page-route.js'
import { nextPage } from '../../routes/next-page.js'
export const buyOrRenew = {
  buy: 'buy',
  renew: 'renew'
}

export const validator = Joi.object({
  'buy-or-renew': Joi.string()
    .valid('buy-licence', 'renew-licence')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

const route = pageRoute(BUY_OR_RENEW.page, BUY_OR_RENEW.uri, validator, nextPage)

export default route
