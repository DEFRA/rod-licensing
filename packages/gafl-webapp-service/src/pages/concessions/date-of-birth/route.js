import { DATE_OF_BIRTH } from '../../../uri.js'
import Joi from 'joi'
import pageRoute from '../../../routes/page-route.js'
import { validation } from '@defra-fish/business-rules-lib'
import { nextPage } from '../../../routes/next-page.js'

export const validator = payload => {
  const dateOfBirth = `${payload['date-of-birth-year']}-${payload['date-of-birth-month']}-${payload['date-of-birth-day']}`
  Joi.assert(
    { 'date-of-birth': dateOfBirth },
    Joi.object({
      'date-of-birth': validation.contact.createBirthDateValidator(Joi)
    })
  )
}

export const getData = async request => {
  const { isLicenceForYou } = await request.cache().helpers.status.getCurrentPermission()

  return { isLicenceForYou }
}

export default pageRoute(DATE_OF_BIRTH.page, DATE_OF_BIRTH.uri, validator, nextPage, getData)
