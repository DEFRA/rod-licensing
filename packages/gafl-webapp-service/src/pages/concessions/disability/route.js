import { DISABILITY_CONCESSION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import * as concessionHelper from '../../../processors/concession-helper.js'
import { nextPage } from '../../../routes/next-page.js'
import { disabilityConcessionTypes } from './update-transaction.js'

const validator = Joi.object({
  'disability-concession': Joi.string()
    .valid(...Object.values(disabilityConcessionTypes))
    .required(),
  'ni-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.pipDla,
    then: validation.contact.createNationalInsuranceNumberValidator(Joi).required(),
    otherwise: Joi.string().empty('')
  }),
  'blue-badge-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.blueBadge,
    then: Joi.string()
      .max(25)
      .required(),
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    hasJunior: concessionHelper.hasJunior(permission),
    hasSenior: concessionHelper.hasSenior(permission),
    isLicenceForYou: permission.isLicenceForYou,
    ...disabilityConcessionTypes
  }
}

export default pageRoute(DISABILITY_CONCESSION.page, DISABILITY_CONCESSION.uri, validator, nextPage, getData)
