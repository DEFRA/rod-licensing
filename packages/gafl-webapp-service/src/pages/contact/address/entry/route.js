import { ADDRESS_ENTRY, ADDRESS_LOOKUP } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { countries } from '../../../../processors/refdata-helper.js'
import { nextPage } from '../../../../routes/next-page.js'
import { getPronoun } from '../../../../processors/licence-type-display.js'

export const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  street: validation.contact.createStreetValidator(Joi),
  locality: validation.contact.createLocalityValidator(Joi),
  town: validation.contact.createTownValidator(Joi),
  postcode: Joi.alternatives().conditional('country-code', {
    is: Joi.string().pattern(/^GB/),
    then: validation.contact.createUKPostcodeValidator(Joi),
    otherwise: validation.contact.createOverseasPostcodeValidator(Joi)
  }),
  'country-code': Joi.string().required()
}).options({ abortEarly: false, allowUnknown: true })

export const getCountryDropDownOptions = async () => {
  const options = await countries.getAll()
  return options.filter(country => country.code !== 'GB')
}

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  const pronoun = getPronoun(isLicenceForYou).possessive

  return {
    pronoun,
    searchTerms: !addresses?.length && searchTerms ? searchTerms : null,
    countries: await getCountryDropDownOptions(),
    lookupPage: ADDRESS_LOOKUP.uri
  }
}

export default pageRoute(ADDRESS_ENTRY.page, ADDRESS_ENTRY.uri, validator, nextPage, getData)
