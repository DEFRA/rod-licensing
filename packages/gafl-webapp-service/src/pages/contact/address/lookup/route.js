import { ADDRESS_LOOKUP, ADDRESS_ENTRY, OS_TERMS } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { hasJunior, validation } from '@defra-fish/business-rules-lib'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  postcode: validation.contact.createUKPostcodeValidator(Joi)
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou: permission.isLicenceForYou,
    licenceLength: permission.licenceLength,
    junior: hasJunior(permission),
    isPhysical: isPhysical(permission),
    uri: {
      entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri),
      osTerms: addLanguageCodeToUri(request, OS_TERMS.uri)
    }
  }
}

export default pageRoute(ADDRESS_LOOKUP.page, ADDRESS_LOOKUP.uri, validator, nextPage, getData)
