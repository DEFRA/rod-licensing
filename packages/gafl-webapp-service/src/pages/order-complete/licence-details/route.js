import Boom from '@hapi/boom'

import pageRoute from '../../../routes/page-route.js'
import { LICENCE_DETAILS } from '../../../uri.js'
import { COMPLETION_STATUS } from '../../../constants.js'
import { nextPage } from '../../../routes/next-page.js'
import { licenceTypeDisplay } from '../../../processors/licence-type-display.js'
import { displayStartTime, displayEndTime } from '../../../processors/date-and-time-display.js'
import * as concessionHelper from '../../../processors/concession-helper.js'

const getAgeConcessionText = (permission, catalog) => {
  if (concessionHelper.hasSenior(permission)) {
    return catalog.age_senior_concession
  }
  if (concessionHelper.hasJunior(permission)) {
    return catalog.age_junior_concession
  }
}

export const getData = async request => {
  const status = await request.cache().helpers.status.get()

  if (!status[COMPLETION_STATUS.agreed]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no agreed flag set')
  }

  if (!status[COMPLETION_STATUS.posted]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no posted flag set')
  }

  if (!status[COMPLETION_STATUS.finalised]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no finalised flag set')
  }

  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  const startTimeString = displayStartTime(request, permission)
  const endTimeString = displayEndTime(request, permission)
  const catalog = request.i18n.getCatalog()

  return {
    permission,
    startTimeString,
    endTimeString,
    disabled: concessionHelper.hasDisabled(permission),
    ageConcession: concessionHelper.getAgeConcession(permission),
    ageConcessionText: getAgeConcessionText(permission, catalog),
    licenceTypeStr: licenceTypeDisplay(permission, catalog)
  }
}

export default pageRoute(LICENCE_DETAILS.page, LICENCE_DETAILS.uri, null, nextPage, getData)
