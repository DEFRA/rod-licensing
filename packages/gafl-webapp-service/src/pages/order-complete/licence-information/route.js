import Boom from '@hapi/boom'

import pageRoute from '../../../routes/page-route.js'
import { LICENCE_INFORMATION } from '../../../uri.js'
import { COMPLETION_STATUS } from '../../../constants.js'
import { nextPage } from '../../../routes/next-page.js'
import { licenceTypeDisplay } from '../../../processors/licence-type-display.js'
import { displayStartTime, displayEndTime } from '../../../processors/date-and-time-display.js'
import * as concessionHelper from '../../../processors/concession-helper.js'

const getData = async request => {
  const status = await request.cache().helpers.status.get()

  if (!status[COMPLETION_STATUS.agreed]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no agreed flag set')
  }

  if (!status[COMPLETION_STATUS.posted]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no posted flag set')
  }

  if (!status[COMPLETION_STATUS.finalised]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no finalised flag set')
  }

  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  const startTimeString = displayStartTime(permission)
  const endTimeString = displayEndTime(permission)

  return {
    permission,
    startTimeString,
    endTimeString,
    disabled: concessionHelper.hasDisabled(permission),
    licenceTypeStr: licenceTypeDisplay(permission)
  }
}

export default pageRoute(LICENCE_INFORMATION.page, LICENCE_INFORMATION.uri, null, nextPage, getData)
