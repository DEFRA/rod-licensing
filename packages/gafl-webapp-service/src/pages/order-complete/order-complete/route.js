import pageRoute from '../../../routes/page-route.js'

import Boom from '@hapi/boom'
import { COMPLETION_STATUS, FEEDBACK_URI_DEFAULT } from '../../../constants.js'
import { ORDER_COMPLETE, NEW_TRANSACTION, LICENCE_INFORMATION } from '../../../uri.js'
import { displayStartTime } from '../../../processors/date-and-time-display.js'
import * as mappings from '../../../processors/mapping-constants.js'
import { nextPage } from '../../../routes/next-page.js'

const getData = async request => {
  const status = await request.cache().helpers.status.get()
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  // If the agreed flag is not set to true then throw an exception
  if (!status[COMPLETION_STATUS.agreed]) {
    throw Boom.forbidden('Attempt to access the completion page handler with no agreed flag set')
  }

  // If the agreed flag is not set to true then throw an exception
  if (!status[COMPLETION_STATUS.posted]) {
    throw Boom.forbidden('Attempt to access the completion page handler with no posted flag set')
  }

  // If the finalised flag has not been set throw an exception
  if (!status[COMPLETION_STATUS.finalised]) {
    throw Boom.forbidden('Attempt to access the completion page handler with no finalised flag set')
  }

  await request.cache().helpers.status.set({ [COMPLETION_STATUS.completed]: true })
  await request.cache().helpers.status.setCurrentPermission({ currentPage: ORDER_COMPLETE.page })

  const startTimeStringTitle = displayStartTime(permission)
  return {
    permission,
    startTimeStringTitle,
    licenceTypes: mappings.LICENCE_TYPE,
    isPostalFulfilment: permission.licensee.postalFulfilment,
    contactMethod: permission.licensee.preferredMethodOfConfirmation,
    howContacted: mappings.HOW_CONTACTED,
    uri: {
      new: NEW_TRANSACTION.uri,
      feedback: process.env.FEEDBACK_URI || FEEDBACK_URI_DEFAULT,
      licenceInformation: LICENCE_INFORMATION.uri
    }
  }
}

export default pageRoute(ORDER_COMPLETE.page, ORDER_COMPLETE.uri, null, nextPage, getData)
