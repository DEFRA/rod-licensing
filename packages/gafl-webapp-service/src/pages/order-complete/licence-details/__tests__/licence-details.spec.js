import { getData } from '../route.js'
import Boom from '@hapi/boom'
import { COMPLETION_STATUS } from '../../../../constants.js'
import { CONCESSION, CONCESSION_PROOF, LICENCE_TYPE } from '../../../../processors/mapping-constants.js'

beforeEach(jest.clearAllMocks)

describe('The licence details page', () => {
  describe('.getData', () => {
    describe('throws a Boom forbidden error', () => {
      it('if status agreed flag is not set', async () => {
        const mockRequest = createMockRequest({})
        const boomError = Boom.forbidden('Attempt to access the licence information handler with no agreed flag set')
        await expect(getData(mockRequest)).rejects.toThrow(boomError)
      })

      it('if status posted flag is not set', async () => {
        const mockRequest = createMockRequest({ [COMPLETION_STATUS.agreed]: true })
        const boomError = Boom.forbidden('Attempt to access the licence information handler with no posted flag set')
        await expect(getData(mockRequest)).rejects.toThrow(boomError)
      })

      it('if status finalised flag is not set', async () => {
        const mockRequest = createMockRequest({ [COMPLETION_STATUS.agreed]: true, [COMPLETION_STATUS.posted]: true })
        const boomError = Boom.forbidden('Attempt to access the licence information handler with no finalised flag set')
        await expect(getData(mockRequest)).rejects.toThrow(boomError)
      })
    })

    it('returns the expected data', async () => {
      const mockRequest = createMockRequest({
        [COMPLETION_STATUS.agreed]: true,
        [COMPLETION_STATUS.posted]: true,
        [COMPLETION_STATUS.finalised]: true
      })
      const result = await getData(mockRequest)
      expect(result).toMatchSnapshot()
    })
  })
})

const createMockRequest = status => ({
  cache: jest.fn(() => ({
    helpers: {
      status: {
        get: jest.fn(() => status)
      },
      transaction: {
        getCurrentPermission: jest.fn(() => ({
          startDate: '2020-06-06',
          endDate: '2021-06-05',
          licenceType: LICENCE_TYPE['trout-and-coarse'],
          numberOfRods: '3',
          concessions: [
            {
              type: CONCESSION.DISABLED,
              proof: {
                type: CONCESSION_PROOF.blueBadge,
                referenceNumber: '123456324'
              }
            },
            {
              type: CONCESSION.SENIOR,
              proof: {
                type: CONCESSION_PROOF.none
              }
            }
          ]
        }))
      }
    }
  })),
  i18n: {
    getCatalog: () => ({
      over_65: 'Over 65',
      licence_type_radio_trout_three_rod: 'Trout and coarse, up to 3 rods',
      renewal_start_date_expires_5: 'on'
    })
  }
})
