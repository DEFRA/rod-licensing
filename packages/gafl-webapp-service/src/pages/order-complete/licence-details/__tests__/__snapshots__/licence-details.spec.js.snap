// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The licence details page .getData returns the expected data 1`] = `
Object {
  "ageConcession": Object {
    "proof": Object {
      "type": "No Proof",
    },
    "type": "Senior",
  },
  "disabled": true,
  "endTimeString": "1:00am on 5 June 2021",
  "licenceTypeStr": "Over 65, Trout and coarse, up to 3 rods",
  "permission": Object {
    "concessions": Array [
      Object {
        "proof": Object {
          "referenceNumber": "123456324",
          "type": "Blue Badge",
        },
        "type": "Disabled",
      },
      Object {
        "proof": Object {
          "type": "No Proof",
        },
        "type": "Senior",
      },
    ],
    "endDate": "2021-06-05",
    "licenceType": "Trout and coarse",
    "numberOfRods": "3",
    "startDate": "2020-06-06",
  },
  "startTimeString": "1:00am on 6 June 2020",
}
`;
